lib/net/ssh/authentication/session.rb

```diff
@@ -67,25 +67,23 @@ def authenticate(next_service, username, password = nil)
           default_keys.each { |key| key_manager.add(key) } unless options.key?(:keys) || options.key?(:key_data)
 
           attempted = []
+          skipped = []
 
           @auth_methods.each do |name|
-            next unless @allowed_auth_methods.include?(name)
-
-            attempted << name
-
-            debug { "trying #{name}" }
-            begin
-              auth_class = Methods.const_get(name.split(/\W+/).map { |p| p.capitalize }.join)
-              method = auth_class.new(self,
-                                      key_manager: key_manager, password_prompt: options[:password_prompt],
-                                      pubkey_algorithms: options[:pubkey_algorithms] || nil)
-            rescue NameError
-              debug {"Mechanism #{name} was requested, but isn't a known type.  Ignoring it."}
+            unless @allowed_auth_methods.include?(name)
+              skipped << name
               next
             end
 
-            return true if method.authenticate(next_service, username, password)
-          rescue Net::SSH::Authentication::DisallowedMethod
+            attempted << name
+            return true if try_authenticate(name, next_service, username, password, key_manager, options)
+          end
+
+          # try skipped methods
```

Comment:
I don't prefer comments, the code is obvious. Or maybe extract to a method like, try_skipped_in_case_allowed_auth_updated, what do you think?